DOCSP-43424 Skip returned results (#133)
mongodb/docs-rust: source/fundamentals/crud/read-operations/skip.txt

.. _rust-skip-guide:

=====================
Skip Returned Results
=====================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: code example, read operation, skip, skip results

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to use the {+driver-long+} to specify the
number of documents to skip in a read operation.

Sample Data for Examples
------------------------

The examples in this guide use the following ``Book`` struct as a model for
documents in the ``books`` collection:

.. literalinclude:: /includes/fundamentals/code-snippets/crud/sort.rs
   :start-after: start-book-struct
   :end-before: end-book-struct
   :language: rust
   :dedent:

The following code shows how to insert sample data into the ``books``
collection:

.. literalinclude:: /includes/fundamentals/code-snippets/crud/sort.rs
   :start-after: start-sample-data
   :end-before: end-sample-data
   :language: rust
   :dedent:

> Skip
> ----

Reviewer: S: [nit] this header could be more descriptive. Maybe "Skip Documents in Operation Results" or just "Skip Documents":
```suggestion
Skip Documents
--------------
```

   :copyable: true

   .. input:: /includes/fundamentals/code-snippets/crud/skip.rs
      :start-after: start-skip-example
      :end-before: end-skip-example
      :language: rust
      :dedent:

   .. output::
      :language: console
      :visible: false

      Book { name: "A Dance with Dragons", author: "Martin", length: 1104 }
      Book { name: "Atlas Shrugged", author: "Rand", length: 1088 }

Options Example
~~~~~~~~~~~~~~~

Alternatively, you can use the ``skip()`` method of the ``FindOptions`` struct.
The following example performs a ``find()`` operation with the following
behavior:

- Sorts the results in descending order on the ``name`` field
- Skips the first document
- Limits the results to two documents
- Returns the remaining documents

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/code-snippets/crud/skip.rs
      :start-after: start-options-skip-example
      :end-before: end-options-skip-example
      :language: rust
      :dedent:

   .. output::
      :language: console
      :visible: false

      Book { name: "Les Misérables", author: "Hugo", length: 1462 }
      Book { name: "Atlas Shrugged", author: "Rand", length: 1088 }

> Aggregation

Reviewer: S: since this is another way of skipping documents, I'd make it a third subsection under "Skip Documents" rather than a separate section

Author: Re-arranged this a little because I didn't want a separate "Example" section to nest, lmk what you think!

-----------

You can use the ``$skip`` stage in an aggregation pipeline to skip documents. To
learn more, see the :ref:`rust-aggregation` guide.

Example
~~~~~~~

> 

Reviewer: S: same suggestions here about "with" and the "Sorts the results..." bullet point

The following example uses the ``aggregate()`` method to perform an aggregation
operation with the following behavior: 

- Sorts the results in ascending order on the ``author`` field 
- Skips the first document
- Returns the remaining documents

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/code-snippets/crud/skip.rs
      :start-after: start-aggregation-example
      :end-before: end-aggregation-example
      :language: rust
      :dedent:

   .. output::
      :language: console
      :visible: false

      Document({"_id": Int32(3), "name": String("Les Misérables"), "author": String("Hugo"), "length": Int32(1462)})
      Document({"_id": Int32(4), "name": String("A Dance with Dragons"), "author": String("Martin"), "length": Int32(1104)})
      Document({"_id": Int32(2), "name": String("Atlas Shrugged"), "author": String("Rand"), "length": Int32(1088)})

Additional Information
----------------------

> To learn more about the operations mentioned in this guide, see the following:

Reviewer: S: add a word describing the links, like "guides" or "resources"
```suggestion
To learn more about the operations mentioned in this guide, see the following guides:
```

- :ref:`rust-query-guide`
- :ref:`rust-retrieve-guide`
- :ref:`rust-compound-operations`
- :ref:`rust-aggregation`
- :ref:`rust-sort-guide`
.. - :ref:`rust-limit-guide`

API Documentation
~~~~~~~~~~~~~~~~~

To learn more about any of the methods or types discussed in this guide, see the

> following API Documentation:

Reviewer: S: you can lowercase "documentation" here
```suggestion
following API documentation:
```

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: code example, read operation, skip, skip results

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

In this guide, you can learn how to use the {+driver-long+} to perform a read
operation that skips a specified number of documents when returning results.

Sample Data for Examples
------------------------

The examples in this guide use the following ``Book`` struct as a model for
documents in the ``books`` collection:

.. literalinclude:: /includes/fundamentals/code-snippets/crud/sort.rs
   :start-after: start-book-struct
   :end-before: end-book-struct
   :language: rust
   :dedent:

The following code shows how to insert sample data into the ``books``
collection:

.. literalinclude:: /includes/fundamentals/code-snippets/crud/sort.rs
   :start-after: start-sample-data
   :end-before: end-sample-data
   :language: rust
   :dedent:

Skip Documents
--------------

This section describes how to skip results retrieved by a query in the following ways:

- :ref:`skip() method <rust-skip-method>`: Chain the ``skip()`` method to the 
  ``find()`` method

> - :ref:`FindOptions::skip() method <rust-findoptions-skip>`: Use the ``skip()``
>   method of the ``FindOptions`` struct

Reviewer: S: Since the FindOptions struct only has one member [method](https://docs.rs/mongodb/3.1.0/mongodb/options/struct.FindOptions.html#method.builder) (`builder()`) and a bunch of [fields](https://docs.rs/mongodb/3.1.0/mongodb/options/struct.FindOptions.html#fields) you can set, I don't know if FindOptions::skip() is the right syntax. Maybe this could just say:
```suggestion
- :ref:`FindOptions struct <rust-findoptions-skip>`: Use the ``skip()`` option builder method to configure a ``FindOptions`` struct
```

> Alternatively, you can :ref:`skip documents in an aggregation pipeline
> <rust-aggregation-skip>`.

Reviewer: S: I'd add this to the list above, since it's a third way to skip results. The bullet could say something like:
```suggestion
- :ref:`Aggregation pipeline <rust-aggregation-skip>`: Create a pipeline that uses the ``$skip`` stage
```

If the number of skipped documents exceeds the number of matched documents for a
query, then that query returns no documents.

Find operations return documents in a natural order that is not sorted on any
fields. To avoid skipping random documents, use the ``sort()`` method to sort
documents on a field with a unique value before setting a skip option. To learn
more, see the :ref:`rust-sort-guide` guide.

.. _rust-skip-method:

skip() Method Example
~~~~~~~~~~~~~~~~~~~~~~

> The ``skip()`` method takes an integer that specifies the number of documents to
> omit from the beginning of the result set. Chain the ``skip()`` method
> to the ``find()`` method to skip documents.

Reviewer: S: I think you can swap the order of these sentences
```suggestion
To skip documents, you can chain the ``skip()`` method
to the ``find()`` method. The ``skip()`` method takes an integer that specifies the number of documents to
omit from the beginning of the result set.
```

This example runs a ``find()`` operation that performs the following actions:

- Sorts the results in ascending order of their ``author`` field values
- Skips the first two documents
- Returns the remaining documents

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/code-snippets/crud/skip.rs
      :start-after: start-skip-example
      :end-before: end-skip-example
      :language: rust
      :dedent:

   .. output::
      :language: console
      :visible: false

      Book { name: "A Dance with Dragons", author: "Martin", length: 1104 }
      Book { name: "Atlas Shrugged", author: "Rand", length: 1088 }

.. _rust-findoptions-skip:

Options Example
~~~~~~~~~~~~~~~

Alternatively, if you are setting and reusing options for your query, you can
use ``FindOptions``. Set the ``skip`` field of the ``FindOptions`` struct by

> using the ``skip()`` setter method. Then, chain the ``with_options()`` method to

Reviewer: S: I suggested saying setter method before (my bad) but based on how [other pages](https://www.mongodb.com/docs/drivers/rust/current/fundamentals/crud/read-operations/cursor/#specify-cursor-behavior) describe these methods, I'd say "option builder method" instead:
```suggestion
using the ``skip()`` option builder method. Then, chain the ``with_options()`` method to
```

the ``find()`` method and pass your ``FindOptions`` struct as a parameter to the
``with_options()`` method.

This example runs a ``find()`` operation that performs the following actions:

- Sorts the results in descending order of their ``name`` field values
- Skips the first document

> - Limits the results to two documents

Reviewer: S: up to you, but since the focus of this page is skip() I don't think you need to include limit() in the example.

- Returns the remaining documents

.. io-code-block::
   :copyable: true

   .. input:: /includes/fundamentals/code-snippets/crud/skip.rs
      :start-after: start-options-skip-example
      :end-before: end-options-skip-example
      :language: rust
      :dedent:

   .. output::
      :language: console
      :visible: false

      Book { name: "Les Misérables", author: "Hugo", length: 1462 }
      Book { name: "Atlas Shrugged", author: "Rand", length: 1088 }

.. _rust-aggregation-skip:

Aggregation Example
~~~~~~~~~~~~~~~~~~~

You can use the ``$skip`` stage in an aggregation pipeline to skip documents. To

> learn more, see the :ref:`rust-aggregation` guide.

Reviewer: S:
```suggestion
learn more about aggregation operations, see the :ref:`rust-aggregation` guide.
```
